### PR TITLE
Migrating the packaging system to `poetry` with `pyproject.toml`

### PR DESCRIPTION
**Pull request by fredgrub**
- Deleted old setup files `requirements.txt`, `setup.cfg`, `setup.py`, `MANIFEST.in`
- Added poetry files `poetry.toml`, `pyproject.toml`, `poetry.lock`
- Added `.pyenv-version` and `.tool-versions` for `pyenv` and `asdf` (useful for testing multiple python versions with `tox`)
- Updated `Makefile`, `CONTRIBUTING.rst`, `tox.ini`

I tested the upload to **PyPI** using **TestPyPI**. The package is located at https://test.pypi.org/project/pipreqs-build-test/ and can be installed via `pip`. I installed it as a user and it worked correctly.

### CODE REVIEW

**Review by Fernando-crz — CHANGES_REQUESTED** [review]
First of all, I had some problems running `poetry install --with dev`.
Got the error "`[Errno 2] No such file or directory: 'python'`". Seems to be an alias problem though, because I could fix the error running
`sudo apt install python-is-python3`
It seems like poetry doesn't support the python command being named "`python3`". Maybe we can fix it using some specific poetry config during installation?

Also, for some reason, I had some problems running the _tox's_ tests in _python version 3.11_. Here's the errors that I got:
```
======================================================================
FAIL: test_get_use_local_only (tests.test_pipreqs.TestPipreqs.test_get_use_local_only)
Test without checking PyPI, check to see if names of local
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fernando/home/labxp/pipreqsxp/pipreqs/tests/test_pipreqs.py", line 142, in test_get_use_local_only
    self.assertTrue(item["name"].lower() in self.local)
AssertionError: False is not true

======================================================================
FAIL: test_init (tests.test_pipreqs.TestPipreqs.test_init)
Test that all modules we will test upon are in requirements file
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fernando/home/labxp/pipreqsxp/pipreqs/tests/test_pipreqs.py", line 169, in test_init
    self.assertEqual(data, sorted(data))
AssertionError: Lists differ: ['ana[186 chars]s==3.1.0', 'pyflakes==3.0.1', 'requests==2.31.[36 chars]4.0'] != ['ana[186 chars]s==3.0.1', 'pyflakes==3.1.0', 'requests==2.31.[36 chars]4.0']

First differing element 9:
'pyflakes==3.1.0'
'pyflakes==3.0.1'

  ['analytics_python==1.4.post1',
   'beautifulsoup4==4.12.2',
   'boto==2.49.0',
   'docopt==0.6.2',
   'flask==2.2.2',
   'flask_seasurf==1.1.1',
   'ipython==8.12.0',
   'nose==1.3.7',
   'peewee==3.17.0',
+  'pyflakes==3.0.1',
   'pyflakes==3.1.0',
-  'pyflakes==3.0.1',
   'requests==2.31.0',
   'sqlalchemy==1.4.39',
   'ujson==5.4.0']

======================================================================
FAIL: test_init_local_only (tests.test_pipreqs.TestPipreqs.test_init_local_only)
Test that items listed in requirements.text are the same
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/fernando/home/labxp/pipreqsxp/pipreqs/tests/test_pipreqs.py", line 195, in test_init_local_only
    self.assertTrue(item[0].lower() in self.local)
AssertionError: False is not true

----------------------------------------------------------------------
```
For some odd reason, those errors only occur running python 3.11, but work on other python versions. Kinda odd because the tests ran in github actions seem to work successfully.

**Comment by mateuslatrova** [comment]
the correct name of the file with the python versions for pyenv is ".python-version"

**Comment by mateuslatrova** [comment]
> First of all, I had some problems running `poetry install --with dev`. Got the error "`[Errno 2] No such file or directory: 'python'`". Seems to be an alias problem though, because I could fix the error running `sudo apt install python-is-python3` It seems like poetry doesn't support the python command being named "`python3`". Maybe we can fix it using some specific poetry config during installation?
> 
> Also, for some reason, I had some problems running the _tox's_ tests in _python version 3.11_. Here's the errors that I got:
> 
> ```
> ======================================================================
> FAIL: test_get_use_local_only (tests.test_pipreqs.TestPipreqs.test_get_use_local_only)
> Test without checking PyPI, check to see if names of local
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/fernando/home/labxp/pipreqsxp/pipreqs/tests/test_pipreqs.py", line 142, in test_get_use_local_only
>     self.assertTrue(item["name"].lower() in self.local)
> AssertionError: False is not true
> 
> ======================================================================
> FAIL: test_init (tests.test_pipreqs.TestPipreqs.test_init)
> Test that all modules we will test upon are in requirements file
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/fernando/home/labxp/pipreqsxp/pipreqs/tests/test_pipreqs.py", line 169, in test_init
>     self.assertEqual(data, sorted(data))
> AssertionError: Lists differ: ['ana[186 chars]s==3.1.0', 'pyflakes==3.0.1', 'requests==2.31.[36 chars]4.0'] != ['ana[186 chars]s==3.0.1', 'pyflakes==3.1.0', 'requests==2.31.[36 chars]4.0']
> 
> First differing element 9:
> 'pyflakes==3.1.0'
> 'pyflakes==3.0.1'
> 
>   ['analytics_python==1.4.post1',
>    'beautifulsoup4==4.12.2',
>    'boto==2.49.0',
>    'docopt==0.6.2',
>    'flask==2.2.2',
>    'flask_seasurf==1.1.1',
>    'ipython==8.12.0',
>    'nose==1.3.7',
>    'peewee==3.17.0',
> +  'pyflakes==3.0.1',
>    'pyflakes==3.1.0',
> -  'pyflakes==3.0.1',
>    'requests==2.31.0',
>    'sqlalchemy==1.4.39',
>    'ujson==5.4.0']
> 
> ======================================================================
> FAIL: test_init_local_only (tests.test_pipreqs.TestPipreqs.test_init_local_only)
> Test that items listed in requirements.text are the same
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/fernando/home/labxp/pipreqsxp/pipreqs/tests/test_pipreqs.py", line 195, in test_init_local_only
>     self.assertTrue(item[0].lower() in self.local)
> AssertionError: False is not true
> 
> ----------------------------------------------------------------------
> ```
> 
> For some odd reason, those errors only occur running python 3.11, but work on other python versions. Kinda odd because the tests ran in github actions seem to work successfully.

I have run in my computer and the installation worked normally

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -76,11 +76,11 @@ Ready to contribute? Here's how to set up `pipreqs` for local development.
...
-    $ flake8 pipreqs tests
-    $ python setup.py test
-    $ tox
-
-   To get flake8 and tox, just pip install them into your virtualenv.
+    $ poetry run flake8 pipreqs tests
+    $ poetry run python -m unittest discover
+    $ poetry run tox
+    
+    To test all versions of python using tox you need to have them installed and for this two options are recommended: `pyenv` or `asdf`.
```
[the remark this answers is not in the document]
Comment (fredgrub) [reply CONTRIBUTING.rst:82]: This is related to the `sphinx' documentation`. To check the final result, run `make docs' and open `docs/_build/html/index.html` and go to the contributing page.

File: tox.ini [hunk 2]
```diff
@@ -1,4 +1,5 @@
 [tox]
+isolated_build = true
```
Comment (mateuslatrova) [inline tox.ini:2]: why is this line necessary now?

File: .python-version [hunk 3]
```diff
@@ -0,0 +1,5 @@
+3.8.18
```
Comment (alan-barzilay) [inline .python-version:1]: is there a need to mention the micro version? couldn' t it be just 3.X instead of 3.X.Y?

File: .tool-versions [hunk 4]
```diff
@@ -0,0 +1 @@
+python 3.8.18 3.9.18 3.10.13 3.11.6 pypy3.9-7.3.12
```
Comment (alan-barzilay) [inline .tool-versions:1]: same question as in python-version, is there a need to mention the micro version? couldn' t it be just 3.X instead of 3.X.Y?

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -99,7 +98,7 @@ Before you submit a pull request, check that it meets these guidelines:
 2. If the pull request adds functionality, the docs should be updated. Put
    your new functionality into a function with a docstring, and add the
    feature to the list in README.rst.
-3. The pull request should work for Python 3.7 to 3.11, and PyPy. Check
+3. The pull request should work for Python 3.8 to 3.11, and PyPy. Check
```
Comment (alan-barzilay) [inline CONTRIBUTING.rst:101]: This type of thing will probably fall between the cracks and get outdated, what if we change it to something like currently supported pyhton versions , and PyPy?

File: poetry.toml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+[virtualenvs]
+prefer-active-python = true
```
Comment (alan-barzilay) [inline poetry.toml:2]: what does this option do exactly?

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    "Natural Language :: English",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11"
+]
+
+[tool.poetry.scripts]
+pipreqs = "pipreqs.pipreqs:main"
```
Comment (alan-barzilay) [inline pyproject.toml:24]: what does this do exactly? is it doing something like setting pipreqs as a cli callable script?

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<3.12"
```
Comment (alan-barzilay) [inline pyproject.toml:27]: is the <3.12 because of that bug or because right now we only claim to support till 3.11?

File: tox.ini (hunk 2 above)
Comment (alan-barzilay) [reply tox.ini:2]: what was it responsible for originally? not sure i understood [the documentation](https://tox.wiki/en/3.3.0/config.html#confval-isolated_build=True|False(default))

It creates a new fresh environment to install the package and test it but now poetry does this for us so keeping it would generate a fresh env inside a fresh env, did I get it right @mateuslatrova ?

File: tox.ini [hunk 9]
```diff
@@ -12,9 +13,7 @@ python =
 [testenv]
 setenv =
     PYTHONPATH = {toxinidir}:{toxinidir}/pipreqs
-commands = python setup.py test
-deps =
```
Comment (alan-barzilay) [inline tox.ini:16]: why are the dependencies not necessary anymore? not a complain, just trying to understand the details of the change

**Review by alan-barzilay — CHANGES_REQUESTED** [review]
Great work @fredgrub and thank you very much @Fernando-crz  and @mateuslatrova for your reviews!

I still haven' t tried to install it locally and test it, I have simply read the changes so I will be back later to let you know if anything goes wrong and if by any chance I can replicate the issue encountered by fernando

File: .python-version (hunk 3 above)
Comment (fredgrub) [reply .python-version:1]: I found out that `pyenv` can auto-resolve the latest versions of each python version, see https://github.com/pyenv/pyenv#prefix-auto-resolution-to-the-latest-version. Indeed, it will no longer be necessary to specify the minor version.

File: .tool-versions (hunk 4 above)
Comment (fredgrub) [reply .tool-versions:1]: Answered at https://github.com/bndr/pipreqs/pull/409#discussion_r1383328090.

File: CONTRIBUTING.rst (hunk 5 above)
Comment (fredgrub) [reply CONTRIBUTING.rst:101]: I agree. This should be more future-proof.

File: poetry.toml (hunk 6 above)
Comment (fredgrub) [reply poetry.toml:2]: Poetry can be installed in several ways and each one associates poetry with a specific version of python. When we run `poetry install`, poetry creates a virtual environment using the version of python it knows. 

This line makes poetry ignore its own version of python and use the version that is active at the moment to create the virtual environment.

This makes managing multiple versions of python easier.

File: pyproject.toml (hunk 7 above)
Comment (fredgrub) [reply pyproject.toml:24]: Sort of. This was inherited from 'setup.py' file (see https://github.com/bndr/pipreqs/pull/409/files?short_path=60f61ab#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L52-L56)

File: pyproject.toml (hunk 8 above)
Comment (fredgrub) [reply pyproject.toml:27]: Supporting python 3.12 affects other files that are not relevant to this PR. It would be better to do it in a separate PR (spoiler: we are working on).

File: tox.ini (hunk 9 above)
Comment (fredgrub) [reply tox.ini:16]: To declare the dependencies, we are using `pyproject.toml`. By setting `isolated_build = true` this will install them during the build step by `tox`.

**Comment by alan-barzilay** [comment]
problem installing in a fresh venv from the test repo:

```
❯ python -m venv teste1
❯ source teste1/bin/activate
❯ pip install -i https://test.pypi.org/simple/ pipreqs-build-test
Looking in indexes: https://test.pypi.org/simple/
Collecting pipreqs-build-test
  Obtaining dependency information for pipreqs-build-test from https://test-files.pythonhosted.org/packages/07/34/71313e66821e8c291c322079e4110fe8e06aca6a97d3343b1e401a8be7f7/pipreqs_build_test-0.4.15-py3-none-any.whl.metadata
  Downloading https://test-files.pythonhosted.org/packages/07/34/71313e66821e8c291c322079e4110fe8e06aca6a97d3343b1e401a8be7f7/pipreqs_build_test-0.4.15-py3-none-any.whl.metadata (7.5 kB)
Collecting docopt==0.6.2 (from pipreqs-build-test)
  Downloading https://test-files.pythonhosted.org/packages/f9/d7/d8f8d3e2d51df48c181d8e3706396c6fc2759b120b6be06721f00d84fe94/docopt-0.6.2-py2.py3-none-any.whl (19 kB)
INFO: pip is looking at multiple versions of pipreqs-build-test to determine which version is compatible with other requirements. This could take a while.
Collecting pipreqs-build-test
  Obtaining dependency information for pipreqs-build-test from https://test-files.pythonhosted.org/packages/07/ea/b8f7be3ae887aaac15343b04a6dd7fec45ef1e1b7dc8dc66e5b093611957/pipreqs_build_test-0.4.14-py3-none-any.whl.metadata
  Downloading https://test-files.pythonhosted.org/packages/07/ea/b8f7be3ae887aaac15343b04a6dd7fec45ef1e1b7dc8dc66e5b093611957/pipreqs_build_test-0.4.14-py3-none-any.whl.metadata (7.5 kB)
  Obtaining dependency information for pipreqs-build-test from https://test-files.pythonhosted.org/packages/1f/e8/04e431dbfd6ee635c5fcf1d7cb0eb90a1af836ecdfb155e5f76a10278063/pipreqs_build_test-0.4.13.1-py3-none-any.whl.metadata
  Downloading https://test-files.pythonhosted.org/packages/1f/e8/04e431dbfd6ee635c5fcf1d7cb0eb90a1af836ecdfb155e5f76a10278063/pipreqs_build_test-0.4.13.1-py3-none-any.whl.metadata (7.5 kB)
ERROR: Cannot install pipreqs-build-test==0.4.13.1, pipreqs-build-test==0.4.14 and pipreqs-build-test==0.4.15 because these package versions have conflicting dependencies.

The conflict is caused by:
    pipreqs-build-test 0.4.15 depends on yarg==0.1.9
    pipreqs-build-test 0.4.14 depends on yarg==0.1.9
    pipreqs-build-test 0.4.13.1 depends on yarg==0.1.9

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts

```

**Comment by alan-barzilay** [comment]
could not replicate the issue encountered by fernando, `poetry install --with dev` worked as intended for me

**Comment by fredgrub** [comment]
@alan-barzilay, answering your comment https://github.com/bndr/pipreqs/pull/409#issuecomment-1795651753, there is a peculiarity regarding using the test-pypi repo. Only `pipreqs-build-test` is available from test-pypi, the other dependencies such as `yarg` aren't guaranteed to be there. So we should explicitly make it fetch the main pypi repo. This is the command to do that
```bash
pip install --extra-index-url https://pypi.org/simple -i https://test.pypi.org/simple/ pipreqs-build-test
```
